Bio/SeqIO/AbiIO.py

```diff
@@ -33,23 +33,250 @@
 # each key is tag_name + tag_number
 # if a tag entry needs to be added, just add its key and its key
 # for the annotations dictionary as the value
+# dictionary for tags that require preprocessing before use in creating
+# seqrecords
 _EXTRACT = {
     'TUBE1': 'sample_well',
     'DySN1': 'dye',
     'GTyp1': 'polymer',
     'MODL1': 'machine_model',
 }
-# dictionary for tags that require preprocessing before use in creating
-# seqrecords
-_SPCTAGS = [
-    'PBAS2',    # base-called sequence
-    'PCON2',    # quality values of base-called sequence
-    'SMPL1',    # sample id inputted before sequencing run
-    'RUND1',    # run start date
-    'RUND2',    # run finish date
-    'RUNT1',    # run start time
-    'RUNT2',    # run finish time
-]
+
+
+# Complete data structure representing 98% of the API. The general section
+# represents the part of the API that's common to ALL instruments, whereas the
+# instrument specific sections are labelled as they are in the ABIF spec
+#
+# The 2%:
+#  - on some fields, there are repeated fields (e.g DYEN1-100), and those are not supported
+#  - OvrI1-N containing color data of some sort
+#
+# Keys don't seem to clash from machine to machine, so when we parse, we look
+# for ANY key, and store that in the raw ABIF data structure attached to the
+# annotations, with the assumption that anyone parsing the data can look up the
+# spec themself
+#
+# Key definitions are retained in case end users want "nice" labels pre-made
+# for them for all of the available fields.
+_INSTRUMENT_SPECIFIC_TAGS = {
+    'general': {
+        'APFN2': 'Sequencing Analysis parameters file name',
+        'APXV1': 'Analysis Protocol XML schema version',
+        'APrN1': 'Analysis Protocol settings name',
+        'APrV1': 'Analysis Protocol settings version',
+        'APrX1': 'Analysis Protocol XML string',
+        'CMNT1': 'Sample Comment',
+        'CTID1': 'Container Identifier, a.k.a. plate barcode',
+        'CTNM1': 'Container name, usually identical to CTID, but not necessarily so',
+        'CTTL1': 'Comment Title',
+        'CpEP1': 'Capillary type electrophoresis. 1 for a capillary based machine. 0 for a slab gel based machine.',
+        'DATA1': 'Channel 1 raw data',
+        'DATA1': 'Channel 1 raw data',
```

Comment:
This looks great! I apologize for my previous comments, I had not realized
that this was specifically what you were doing (PLOC1 is the tag for my
instrument, which you have already added) :) This is a feature I have been
hoping someone would implement. Thank you!

On Thu, Feb 12, 2015 at 10:25 AM, Eric Rasche <notifications@github.com>
wrote:

> In Bio/SeqIO/AbiIO.py
> <https://github.com/biopython/biopython/pull/473#discussion_r24588266>:
>
> > +# Key definitions are retained in case end users want "nice" labels pre-made
> > +# for them for all of the available fields.
> > +_INSTRUMENT_SPECIFIC_TAGS = {
> > +    'general': {
> > +        'APFN2': 'Sequencing Analysis parameters file name',
> > +        'APXV1': 'Analysis Protocol XML schema version',
> > +        'APrN1': 'Analysis Protocol settings name',
> > +        'APrV1': 'Analysis Protocol settings version',
> > +        'APrX1': 'Analysis Protocol XML string',
> > +        'CMNT1': 'Sample Comment',
> > +        'CTID1': 'Container Identifier, a.k.a. plate barcode',
> > +        'CTNM1': 'Container name, usually identical to CTID, but not necessarily so',
> > +        'CTTL1': 'Comment Title',
> > +        'CpEP1': 'Capillary type electrophoresis. 1 for a capillary based machine. 0 for a slab gel based machine.',
> > +        'DATA1': 'Channel 1 raw data',
> > +        'DATA1': 'Channel 1 raw data',
>
> (Maybe I'm still not clear on your comment, if so, my apologies in
> advance).
>
> Instrument ID seems to be pretty reliably specified in MODL1 so that can
> be extracted.
>
> However, as I was building the _INSTRUMENT_SPECIFIC_TAGS dictionary, I
> didn't notice a single case where multiple (different) keys on different
> instruments mapped to the same data. All of the keys in the general set I
> checked the definitions of, to ensure they were the same for each
> instrument, before merging them into the general section.
>
> peak position data, which has different tags for different instruments
>
> Searching through the official spec for all instances of the word peak,
> I'm really not seeing anything on instruments other than the 3500. If that
> information is tagged differently in ABIF files from different instruments,
> it's not in the spec. If you have examples I'd be happy to add restructure
> the code to accomodate that.
>
> The data structure in my commit was largely generated from a pdftotext
> rendering of the PDF file, followed by some transforms and then manually
> merging the keys from each instruments if they had identical definitions.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/biopython/biopython/pull/473/files#r24588266>.
>